backend/app/datamodels/tabs.js

```diff
@@ -1,35 +1,95 @@
-const mongoose = require('mongoose');
 const { Tabs } = require('../db/models');
 
+const getTab = async (tabId) => {
+	let tab = await Tabs.findById(tabId);
+	if (!tab) {
+		throw new Error('Tab not found');
+	} else {
+		tab = await tab.populate('orders').execPopulate();
+		return tab;
+	}
+};
+
 const closeTab = async (tabId, session) => {
-    const tab = await Tabs.findById(tabId).session(session);
+	const tab = await Tabs.findById(tabId).session(session);
+
+	if (!tab) {
+		throw new Error('Tab not found');
+	}
 
-    if (!tab) {
-        throw new Error('Tab not found');
-    }
+	tab.status = 'closed';
+	tab.lastUpdated = Date.now();
 
-    tab.status = 'closed';
-    tab.lastUpdated = Date.now();
+	await tab.save({ session });
 
-    await tab.save({ session });
+	return tab;
+};
 
-    return tab;
-}
+const clearWaiterRequested = async (tabId) => {
+	const tab = await Tabs.findById(tabId);
 
-const clearWaiter = async (tabId) => {
-    const tab = await Tabs.findById(tabId);
+	if (!tab) {
+		throw new Error('Tab not found');
+	}
 
-    if (!tab) {
-        throw new Error('Tab not found');
-    }
+	tab.callWaiter = null;
+	await tab.save();
 
-    tab.callWaiter = null;
-    await tab.save();
+	return tab;
+};
 
-    return tab;
-}
+const waiterRequested = async (tabId) => {
+	const tab = await Tabs.findById(tabId);
+	if (!tab) {
+		throw new Error('Tab not found');
+	}
+	if (tab.callWaiter === 'called') {
+		throw new Error('Waiter already called');
+	}
+	if (tab.status === 'closed') {
+		throw new Error('Tab closed');
+	} else {
+		tab.callWaiter = 'called';
+		await tab.save();
+	
+		return true;
+	}
+};
+
+const paymentRequested = async (tabId, paymentType, tipsPercent) => {
+	const tab = await Tabs.findById(tabId);
+	if (!tab) {
+		throw new Error('Tab not found');
+	}
+	if (tab.status === 'closed') {
+		throw new Error('Tab closed');
+	}
+	
+	tab.paymentRequested = paymentType;
+	if (tipsPercent) {
+		tab.tipsPercent = tipsPercent;
+	}
+	await tab.save();
+
+	return true;
+};
+
+const newTab = async (table) => {
+	const newTab = await new Tabs({
+		restaurantId: table.restaurantId,
+		tableId: table._id,
+		status: 'open',
+		lastUpdated: Date.now(),
+		createdAt: Date.now()
+	}).save();
+	return newTab;
+};
 
 module.exports = {
-    closeTab,
-    clearWaiter,
+	getTab,
+	closeTab,
+	newTab,
+	waiterRequested,
+	clearWaiterRequested,
+	paymentRequested
 };
```

Comment:
The code patch seems to be a set of functions related to managing restaurant tabs. Here are some suggestions based on the code:

1. It is good practice to declare `const mongoose = require('mongoose');` at the beginning of the file, since it is used in multiple functions.

2. The `getTab` function is improved by first checking if the tab exists and then proceeding to populate its orders. This will avoid unnecessary database queries.

3. In the `closeTab`, `clearWaiterRequested`, `waiterRequested`, and `paymentRequested` functions, it is important to validate that the tab exists before modifying it. Otherwise, errors may occur.

4. The `waiterRequested` function could also include a check for whether the table has already requested payment or not.

5. Finally, it is recommended to use more descriptive variable names. For example, instead of `tab` use `currentTab`, or similar. This will make it easier to read and maintain the code.

Overall, the code looks clean, and adding error handling is essential when dealing with databases.